applications/GeoMechanicsApplication/custom_strategies/strategies/geo_mechanics_newton_raphson_erosion_process_strategy.hpp

```diff
@@ -80,42 +98,31 @@ class GeoMechanicsNewtonRaphsonErosionProcessStrategy
-            return dynamic_cast<SteadyStatePwPipingElement<2, 4>*>(p_element);
-        });
-
-        const auto number_of_piping_interface_elements = static_cast<std::size_t>(
-            std::count_if(piping_interface_elements.begin(), piping_interface_elements.end(),
-                          [](auto p_element) { return p_element != nullptr; }));
-        KRATOS_ERROR_IF(number_of_piping_interface_elements != 0 &&
-                        number_of_piping_interface_elements != piping_interface_elements.size())
-            << "Unexpected number of piping interface elements: expected either 0 or "
-            << piping_interface_elements.size() << " but got "
-            << number_of_piping_interface_elements << std::endl;
-
-        if (number_of_piping_interface_elements > 0) {
-            this->DetermineOpenPipingElements(piping_interface_elements);
+        auto piping_interface_elements =
+            TryDownCastToPipingElement<SteadyStatePwPipingElement<2, 4>>(piping_elements);
+        if (piping_interface_elements) {
+            this->DetermineOpenPipingElements(piping_interface_elements.value());
             this->BaseClassFinalizeSolutionStep();
             return;
         }
 
-        auto piping_line_elements = std::vector<GeoSteadyStatePwPipingElement<2, 2>*>{};
-        std::transform(piping_elements.begin(), piping_elements.end(),
-                       std::back_inserter(piping_line_elements), [](auto p_element) {
-            return dynamic_cast<GeoSteadyStatePwPipingElement<2, 2>*>(p_element);
-        });
+        auto piping_2D_line_elements =
+            TryDownCastToPipingElement<GeoSteadyStatePwPipingElement<2, 2>>(piping_elements);
+        if (piping_2D_line_elements) {
+            this->DetermineOpenPipingElements(piping_2D_line_elements.value());
+            this->BaseClassFinalizeSolutionStep();
+            return;
+        }
 
-        const auto number_of_piping_line_elements = static_cast<std::size_t>(
-            std::count_if(piping_line_elements.begin(), piping_line_elements.end(),
-                          [](auto p_element) { return p_element != nullptr; }));
-        KRATOS_ERROR_IF(number_of_piping_line_elements != piping_line_elements.size())
-            << "Unexpected number of piping line elements: expected " << piping_line_elements.size()
-            << " but got " << number_of_piping_line_elements << std::endl;
+        auto piping_3D_line_elements =
+            TryDownCastToPipingElement<GeoSteadyStatePwPipingElement<3, 2>>(piping_elements);
+        if (piping_3D_line_elements) {
+            this->DetermineOpenPipingElements(piping_3D_line_elements.value());
+            this->BaseClassFinalizeSolutionStep();
+            return;
+        }
 
-        this->DetermineOpenPipingElements(piping_line_elements);
-        this->BaseClassFinalizeSolutionStep();
+        KRATOS_ERROR << "Unexpected type of piping line element\n";
```

Comment:
The message text is correct, as also the interface element is in a way a line element. However, I would prefer a simpeler message: "Unexpected type of piping element\n";
I don't really expect this message to be reached, The types of piping elements are already handled way before this FinalizeSolutionStep.